Datajson v6.6

regit opened the pull request:
Update of #13343 

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [x] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [x] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues

Link to ticket: https://redmine.openinfosecfoundation.org/issues/7372

Describe changes:
- rebase on master
- address scan-build
- fix leak on unix socket addition

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2554

regit commented:
CI green and remarks are mostly implemented. Only missing part is the removal of non related commits.

victorjulien commented:
Passed my QA. Ran this PR with SV PR https://github.com/OISF/suricata-verify/pull/2554. Local pipeline 5421, run 904.

victorjulien commented on src/detect.h:
```diff
@@ -1205,6 +1205,16 @@ typedef struct PostRuleMatchWorkQueue {
+
+/** structure to store the json content with info on sig that triggered it */
+typedef struct SigJsonContent {
+    void *id;
```
field appears to be unused

regit replied:
it is not adding a comment to explain

victorjulien reviewed (commented):
@regit please see inline. Quite a few things to improve still, with the most important one addressing the confusing locking.

I'm willing to take a chance on this and merge it now if you're committing to addressing the issues in the next week or 2.

regit commented:
> @regit please see inline. Quite a few things to improve still, with the most important one addressing the confusing locking.
> 
> I'm willing to take a chance on this and merge it now if you're committing to addressing the issues in the next week or 2.

@victorjulien I will work on addressing the issue today so it should be ready for this week.

victorjulien commented:
> > @regit please see inline. Quite a few things to improve still, with the most important one addressing the confusing locking.
> > I'm willing to take a chance on this and merge it now if you're committing to addressing the issues in the next week or 2.
> 
> @victorjulien I will work on addressing the issues today so it should be ready for this week.

Ok will stage this PR then, so please keep it open.

regit commented on src/detect-dataset.c:
```diff
@@ -203,7 +258,43 @@ static int DetectDatasetParse(const char *str, char *cmd, int cmd_len, char *nam
+                } else if (strcmp(val, "json") == 0) {
+                    *format = DATASET_FORMAT_JSON;
+                } else {
+                    SCLogWarning("unknown format %s", val);
```
Fixing it globally in the function.

regit commented on src/detect-dataset.c:
```diff
@@ -63,7 +64,51 @@ void DetectDatasetRegister (void)
+            // PrintRawDataFp(stdout, data, data_len);
+            DataJsonResultType r = DatajsonLookup(sd->set, data, data_len);
+            SCLogDebug("r found: %d, len: %u", r.found, r.json.len);
+            if (r.found)
```
good catch.

regit commented on src/detect-dataset.c:
```diff
@@ -63,7 +64,51 @@ void DetectDatasetRegister (void)
+                    det_ctx->json_content[det_ctx->json_content_len].id = sd->id;
+                    det_ctx->json_content_len++;
+                }
+                DatajsonUnlockElt(&r);
```
it should indeed be moved outside of conditional.

regit commented:
@victorjulien I've addressed the comments on this branch https://github.com/regit/suricata/tree/datajson-update-v1.0

I'm currently trying it a bit but it passes suricata-verify tests so it is production ready ;)

Let me know if you want a PR.

victorjulien commented:
> @victorjulien I've addressed the comments on this branch https://github.com/regit/suricata/tree/datajson-update-v1.0
> 
> I'm currently trying it a bit but it passes suricata-verify tests so it is production ready ;)
> 
> Let me know if you want a PR.

Staging is currently running, so lets do it as a follow up PR after that is merged.

victorjulien commented:
btw @regit code coverage is really low. Should try to get at least 80% of the diff covered.

victorjulien commented:
> > Needs performance optimizations to be able to get merged. Last 6 QA runs here and in the staging PR all failed to run within the time constraints.
> 
> I got 2 questions:
> 
>     * Do you want me to PR the fixes discussed yesterday ?
> 
>     * Is it possible to find more information or diagnosis about the performance results ?

Ya lets get the latest code.

On perf issues, all I have is "it runs slower in QA" :)

victorjulien commented:
Continued in #13408
